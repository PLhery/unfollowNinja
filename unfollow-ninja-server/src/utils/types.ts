import type { UserCategory } from '../dao/dao';
import { SUPPORTED_LANGUAGES_CONST } from './utils';

export type Lang = typeof SUPPORTED_LANGUAGES_CONST[number];

export interface IFollowerInfo {
    id: string;
    followTime: number;
}

export interface IUnfollowerInfo extends IFollowerInfo {
    unfollowTime: number;
    followDetectedTime: number | null; // first time the system saw the follower
    blocking?: boolean;
    blocked_by?: boolean;
    suspended?: boolean;
    locked?: boolean;
    deleted?: boolean;
    following?: boolean;
    followed_by?: boolean;
    friendship_error_code?: number;
    notified_time?: number;
    username?: string;
    skippedBecauseGlitchy?: boolean;
}

export interface ITwittoInfo {
    id: string;
    username: string;
}

export interface IUserParams {
    added_at: number; // in ms
    lang: Lang;
    token: string;
    tokenSecret: string;
    isTemporarySecondAppToken?: boolean;
    dmId?: string;
    dmToken?: string;
    dmTokenSecret?: string;
    pro?: '3' | '2' | '1' | '0'; // friendcode-friends-pro-normal
    friendCodes?: string;
    customerId?: string;
}

export interface IUserEgg extends ITwittoInfo, IUserParams {
    category: UserCategory;
}

export interface Session {
    user?: IUserEgg & { dmUsername?: string };
}
