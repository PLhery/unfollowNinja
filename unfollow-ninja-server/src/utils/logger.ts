import cluster from 'cluster';
import * as fs from 'fs';
import { createLogger, format, transports } from 'winston';

let workerInfo = cluster.worker ? `work ${cluster.worker.id}` : 'master';
export const setLoggerPrefix = (prefix: string) => workerInfo = prefix + ' '.repeat(6 - prefix.length);

const customFormat = format.combine(
    format.timestamp(),
    format.splat(),
    format.errors({ stack: true }),
    format.printf((info) => `${info.timestamp} ${workerInfo} ${info.level}: ${info.message}`),
);

const fileParams = {
    maxsize: 20000000, // 20MB
    maxFiles: 200,
};

const testEnv = typeof it === 'function'; // jest

fs.exists( './logs', (exists) => !exists && fs.mkdir('./logs', () => null));

const logger = createLogger({
    format: customFormat,
    transports: [
        new transports.Console({
            format: format.combine(format.colorize(), customFormat),
            level: testEnv ? 'warn' : 'info',
        }),
        !testEnv && new transports.File({
            ...fileParams,
            filename: 'logs/info.log',
            level: 'info',
        }),
        !testEnv && new transports.File({
            ...fileParams,
            filename: 'logs/error.log',
            level: 'error',
        }),
        !testEnv && new transports.File({
            ...fileParams,
            filename: 'logs/debug.log',
            level: 'debug',
        }),
        !testEnv && new transports.File({
            ...fileParams,
            filename: 'logs/warn.log',
            level: 'warn',
        }),
    ].filter(t => t),
    exceptionHandlers: [
        new transports.Console(),
        !testEnv && new transports.File({
            ...fileParams,
            filename: 'logs/exceptions.log',
        }),
    ].filter(t => t),
});

export default logger;
