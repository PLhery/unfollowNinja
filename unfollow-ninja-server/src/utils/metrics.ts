import { StatsD } from 'hot-shots';

const STATSD_HOST = process.env.STATSD_HOST || undefined;
const DD_AGENT_HOST = process.env.DD_AGENT_HOST || undefined;
const METRICS_PREFIX = process.env.METRICS_PREFIX || 'uninja';

export default class Metrics {
    private static statsDClients: StatsD[] = [];

    public static addStatsdHost(host: string, port: number = 8125) {
        this.statsDClients.push(new StatsD({host, port}))
    }

    public static gauge(metric: string, value: number) {
        this.statsDClients
            .forEach(client => client.gauge(METRICS_PREFIX + '.' + metric, value));
    }

    public static increment(metric: string, value: number = 1) {
        this.statsDClients
            .forEach(client => client.increment(METRICS_PREFIX + '.' + metric, value));
    }

    public static kill() {
        this.statsDClients.forEach(client => client.close(null))
    }
}
if (STATSD_HOST) {
    Metrics.addStatsdHost(STATSD_HOST);
}
if (DD_AGENT_HOST) { // datadog
    Metrics.addStatsdHost(DD_AGENT_HOST, Number(process.env.DD_DOGSTATSD_PORT || 8125));
}