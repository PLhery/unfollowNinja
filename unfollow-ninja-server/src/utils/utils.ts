// Convert an ID generated with snowflake (e.g some cursors) to a timestamp in ms
import bigInt from 'big-integer';

/**
 * get the timestamp associated with a Twitter's cursor
 */
export function twitterCursorToTime(cursor: string): number {
    return cursor ? bigInt(cursor).abs().shiftRight(20).valueOf() : null;
}
