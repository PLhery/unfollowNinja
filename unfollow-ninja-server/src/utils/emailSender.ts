import 'dotenv/config';
import { google } from 'googleapis';
import MailComposer from 'nodemailer/lib/mail-composer';
import Stripe from 'stripe';
import logger from './logger';

const auth =
    process.env.GOOGLE_CLIENT_ID && process.env.GOOGLE_CLIENT_SECRET && process.env.GOOGLE_REFRESH_TOKEN
        ? new google.auth.OAuth2(process.env.GOOGLE_CLIENT_ID, process.env.GOOGLE_CLIENT_SECRET)
        : null;
if (auth) {
    auth.setCredentials({ refresh_token: process.env.GOOGLE_REFRESH_TOKEN });
}

const client = google.gmail({ version: 'v1', auth });

const stripe = process.env.STRIPE_SK
    ? new Stripe(process.env.STRIPE_SK, {
          apiVersion: '2022-11-15',
      })
    : null;

const encodeMessage = (message) =>
    Buffer.from(message).toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');

export async function sendRevokedEmailToUserId(userId: string) {
    if (!stripe || !client) {
        return;
    }
    const email = await emailFromUserId(userId);
    if (email) {
        logger.info(`${userId} is revoked, sending revoked email to ${email}`);
        await sendRevokedEmail(email);
    }
}

/**
 * Get the Stripe email address of a user from its userId
 * @param userId the userId
 */
async function emailFromUserId(userId: string) {
    const activeSubscription = (
        await stripe.subscriptions.search({
            query: `metadata['userId']:'${userId}'`,
            limit: 100,
            expand: ['data.customer'],
        })
    ).data.filter((s) => s.status === 'active' || s.status === 'trialing')[0];

    return (activeSubscription?.customer as Stripe.Customer | null)?.email;
}

/**
 * Send an email to a user to tell him that his Twitter account has been disconnected from Unfollow Monkey
 * @param to the email address of the user
 */
export async function sendRevokedEmail(to: string) {
    const mailComposer = new MailComposer({
        to,
        from: 'Unfollow Monkey <love@unfollow-monkey.com>',
        subject: 'Your Twitter account has been disconnected from Unfollow Monkey',
        text:
            'Hello,\n' +
            '\n' +
            'It looks like your Unfollow Monkey app has been disconnected from Twitter.\n' +
            '\n' +
            'This can happen when Twitter accounts are temporarily disabled for instance.\n' +
            '\n' +
            'Please\n' +
            "- Log in again to Unfollow Monkey if you'd like to fix this\n" +
            "- or tell me whether you'd like to cancel your subscription\n" +
            '\n' +
            'Thank you and best regards,\n' +
            '\n' +
            'Paul',
        textEncoding: 'base64',
    });
    const message = await mailComposer.compile().build();
    await client.users.messages.send({
        userId: 'me',
        requestBody: {
            raw: encodeMessage(message),
        },
    });
}
