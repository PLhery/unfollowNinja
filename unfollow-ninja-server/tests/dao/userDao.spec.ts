import Redis from 'ioredis';
import {Sequelize} from 'sequelize';
import RedisMock from 'ioredis-mock';

import Dao, {UserCategory} from '../../src/dao/dao';
import {IUserEgg, IUserParams} from '../../src/utils/types';

const redis: Redis = process.env.REDIS_TEST_URI ?
    new Redis(process.env.REDIS_TEST_URI, { lazyConnect: true }) :
    new RedisMock({ lazyConnect: true });

const sequelize = process.env.POSTGRES_TEST_URI ?
  new Sequelize(process.env.POSTGRES_TEST_URI, { logging: false }) :
  new Sequelize( { dialect: 'sqlite', storage: ':memory:', logging: false});

const sequelizeLogs = process.env.POSTGRES_LOGS_TEST_URI ?
  new Sequelize(process.env.POSTGRES_LOGS_TEST_URI, { logging: false }) :
  new Sequelize( { dialect: 'sqlite', storage: ':memory:', logging: false });

const dao = new Dao(redis, sequelize, sequelizeLogs);

const uDao1 = dao.getUserDao( '1');
const uDao2 = dao.getUserDao('2');

const USER_PARAMS_1: IUserParams = {
    added_at: 1234,
    lang: 'fr',
    token: 't0k3n',
    tokenSecret: 's3cr3t'
};
const USER_PARAMS_2: IUserParams = {
    added_at: 2345,
    lang: 'en',
    token: 't0k4n',
    tokenSecret: 's2cr2t',
    dmId: '3',
    dmToken: 'token',
    dmTokenSecret: 'secret',
    pro: '1',
    customerId: 'cus_xXx1fff000999fY'
};

describe('Test userDao', () => {
    afterAll(async () => {
        await redis.flushdb();
        await sequelize.drop();
        await dao.disconnect();
    });

    beforeAll(async () => {
        await sequelize.drop();
        await dao.load();
        await redis.flushdb();

        const user1: IUserEgg = {
            ...USER_PARAMS_1,
            id: '1',
            username: 'user 1',
            category: UserCategory.enabled,
        };
        const user2: IUserEgg = {
            ...USER_PARAMS_2,
            id: '2',
            username: 'user 2',
            category: UserCategory.disabled,
        };
        await dao.addUser(user1);
        await dao.addUser(user2);
    });

    test('should get the usernames', async () => {
        expect(await uDao1.getUsername()).toBe('user 1');
        expect(await uDao2.getUsername()).toBe('user 2');
    });

    test('should be able to read and edit the category', async () => {
        expect(await dao.getUserIdsByCategory(UserCategory.enabled)).toStrictEqual(['1']);
        expect(await dao.getUserIdsByCategory(UserCategory.disabled)).toStrictEqual(['2']);
        expect(await dao.getUserIdsByCategory(UserCategory.revoked)).toStrictEqual([]);
        expect(await uDao1.getCategory()).toBe(UserCategory.enabled);
        expect(await uDao2.getCategory()).toBe(UserCategory.disabled);

        await uDao1.setCategory(UserCategory.revoked);

        expect(await dao.getUserIdsByCategory(UserCategory.enabled)).toStrictEqual([]);
        expect(await dao.getUserIdsByCategory(UserCategory.disabled)).toStrictEqual(['2']);
        expect(await dao.getUserIdsByCategory(UserCategory.revoked)).toStrictEqual(['1']);
        expect(await uDao1.getCategory()).toBe(UserCategory.revoked);
        expect(await uDao2.getCategory()).toBe(UserCategory.disabled);

        await uDao1.enable();
        expect(await uDao1.getCategory()).toBe(UserCategory.enabled);
    });

    test('should be able to save the next time to check', async () => {
        expect(await uDao1.getNextCheckTime()).toBe(0); // defaults to 0
        await uDao1.setNextCheckTime(123);
        await uDao2.setNextCheckTime(234);
        await uDao2.setNextCheckTime(345);
        expect(await uDao1.getNextCheckTime()).toBe(123);
        expect(await uDao2.getNextCheckTime()).toBe(345);
    });

    test('should be able to fetch and edit user params', async () => {
        const uParamsStr1 = {...USER_PARAMS_1,
          added_at: 1234, dmId: '', dmToken: '', dmTokenSecret: '', customerId: '', pro: '0'};
        const uParamsStr2 = {...USER_PARAMS_2, added_at: 2345, pro: '1'};
        expect(await uDao1.getUserParams()).toStrictEqual(uParamsStr1);
        expect(await uDao2.getUserParams()).toStrictEqual(uParamsStr2);

        const newParams = {dmId: '3030', dmToken: 'token2', dmTokenSecret: 'secret2',
            token: 't0k4n2', tokenSecret: 's2cr2t2'};
        await uDao2.setUserParams(newParams);
        expect(await uDao2.getUserParams()).toStrictEqual({...uParamsStr2, ...newParams});
        await uDao2.setUserParams(USER_PARAMS_2);
        expect(await uDao2.getUserParams()).toStrictEqual(uParamsStr2);
    });

    test('should be able to get a Twit instance for each user', async () => {
        process.env.CONSUMER_KEY = 'ckey';
        process.env.CONSUMER_SECRET = 'csecret';
        const twit1 = await uDao1.getTwit();
        const twit2 = await uDao2.getTwit();

        expect(twit1.getAuth()).toStrictEqual({
            consumer_key: 'ckey',
            consumer_secret: 'csecret',
            access_token: USER_PARAMS_1.token,
            access_token_secret: USER_PARAMS_1.tokenSecret,
        });
        expect(twit2.getAuth()).toStrictEqual({
            consumer_key: 'ckey',
            consumer_secret: 'csecret',
            access_token: USER_PARAMS_2.token,
            access_token_secret: USER_PARAMS_2.tokenSecret,
        });
    });

    test('should be able to get a dmTwit instance for each user', async () => {
        process.env.DM_CONSUMER_KEY = 'dmckey';
        process.env.DM_CONSUMER_SECRET = 'dmcsecret';
        const dmTwit2 = await uDao2.getDmTwit();

        expect(dmTwit2.getAuth()).toStrictEqual({
            consumer_key: 'dmckey',
            consumer_secret: 'dmcsecret',
            access_token: USER_PARAMS_2.dmToken,
            access_token_secret: USER_PARAMS_2.dmTokenSecret,
        });
        await expect(uDao1.getDmTwit()).rejects.toThrow('the user didn\'t have any DM credentials stored');
    });

    test('should be able to get the language', async () => {
        expect(await uDao1.getLang()).toBe('fr');
        expect(await uDao2.getLang()).toBe('en');
    });

    test('isPro', async () => {
      expect(await uDao1.isPro()).toBe(false);
      expect(await uDao2.isPro()).toBe(true);
    });

    test('should store and retrieve a list of followers', async () => {
        expect(await uDao1.getFollowers()).toBeNull();

        await uDao1.updateFollowers(['1', '2', '3'], ['1', '2', '3'], [], 123);
        await uDao1.updateFollowers(['1', '4'], ['4'], ['2', '3'], 456);

        expect(await uDao1.getFollowers()).toStrictEqual(['1', '4']);
        expect(await redis.get('followers:count:1')).toBe('2');
        expect(await uDao1.getFollowDetectedTime('1')).toBe(123);
        expect(await uDao1.getFollowDetectedTime('2')).toBe(0);
        expect(await uDao1.getFollowDetectedTime('3')).toBe(0);
        expect(await uDao1.getFollowDetectedTime('4')).toBe(456);
        expect(await uDao1.getFollowTime('4')).toBe(456);
        expect(await redis.get('total-unfollowers')).toBe('2');
    });

    test('should store snowflake IDs', async () => {
        await uDao1.setFollowerSnowflakeId('1', '1654482657084000');
        await uDao1.setFollowerSnowflakeId('4', '1654482657084000');
        expect(await uDao1.getFollowerSnowflakeId('1')).toBe('1654482657084000');
        expect(await uDao1.getCachedFollowers()).toStrictEqual(['1', '4']);
        expect(await uDao1.getFollowTime('1')).toBe(1577837617); // uses snowflake this time

        await uDao1.removeFollowerSnowflakeIds(['1']);
        expect(await uDao1.getFollowerSnowflakeId('1')).toBeNull();
        expect(await uDao1.getCachedFollowers()).toStrictEqual(['4']);
    });

    test('should manage cached/uncached followers', async () => {
        expect(await uDao1.getHasNotCachedFollowers()).toBe(true);

        await uDao1.setFollowerSnowflakeId('4', '1654482657084000');
        expect(await uDao1.getHasNotCachedFollowers()).toBe(true);

        await uDao1.addUncachableFollower('1');
        expect(await uDao1.getHasNotCachedFollowers()).toBe(false);
        expect(await uDao1.getUncachableFollowers()).toStrictEqual(['1']);
    });

    test('should manage friend codes', async () => {
      expect(await uDao1.getFriendCodes()).toHaveLength(0);
      await uDao1.addFriendCodes();
      await uDao2.addFriendCodes();
      await uDao1.addFriendCodes(); // the second call should not do anything
      const codes = await uDao1.getFriendCodes();
      expect(codes).toHaveLength(5);
      expect(codes[0].userId).toBe('1');
      expect(codes[0].code).toHaveLength(6);

      expect(await uDao2.registerFriendCode('AAAAAA')).toBe(false);
      expect(await uDao2.registerFriendCode(codes[1].code)).toBe(true);
      expect(await uDao2.registerFriendCode(codes[1].code)).toBe(false);
      expect((await uDao2.getRegisteredFriendCode())?.userId).toBe('1');

      await uDao2.deleteFriendCodes(codes[1].code);
      expect(await uDao1.getFriendCodes()).toHaveLength(5);
      expect(await uDao2.getFriendCodes()).toHaveLength(5);

      await uDao1.deleteFriendCodes(codes[1].code);
      expect(await uDao1.getFriendCodes()).toHaveLength(4);
      expect(await uDao2.getFriendCodes()).toHaveLength(5);
      expect((await uDao2.getRegisteredFriendCode())).toBe(null);
    });

    // depends heavily on other tests
    test('should get a stable getAllUserData', async () => {
      const data = await uDao1.getAllUserData();
      delete data.friendCodes; // not stable
      delete data.registeredFriendCode; // not stable
      expect(data).toMatchSnapshot();
    });

    test('should completely delete data about the user', async () => {
        expect(await redis.dbsize()).toBe(11);
        await uDao1.deleteUser();
        await uDao2.deleteUser();
        expect(await redis.zcard('users')).toBe(0);
        redis.del('users'); // empty users appears as a key on ioredis-mock but not on actual redis 6
        expect((await redis.keys('*')).sort()).toEqual(['total-unfollowers']);
    });
});
