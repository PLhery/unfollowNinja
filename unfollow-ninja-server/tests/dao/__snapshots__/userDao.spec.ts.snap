// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test userDao should get a stable getAllUserData 1`] = `
Object {
  "category": 0,
  "followTimes": Object {
    "1": "123",
    "4": "456",
  },
  "followers": Array [
    "1",
    "4",
  ],
  "nextCheckTime": 123,
  "snowflakeIds": Object {
    "4": "1654482657084000",
  },
  "uncachables": Array [
    "1",
  ],
  "userParams": Object {
    "added_at": 1234,
    "customerId": "",
    "dmId": "",
    "dmToken": "",
    "dmTokenSecret": "",
    "lang": "fr",
    "pro": "0",
    "token": "t0k3n",
    "tokenSecret": "s3cr3t",
  },
  "username": "user 1",
}
`;
